src/main/java/org/digma/intellij/plugin/toolwindow/recentactivity/RecentActivityService.java

```diff
@@ -1,47 +1,389 @@
+    private static final String RECENT_EXPIRATION_LIMIT_VARIABLE = "recentActivityExpirationLimit";
+    private static final int FETCHING_LOOP_INTERVAL = 10 * 1000; // 10sec
 
+    private final Logger logger = Logger.getInstance(RecentActivityService.class);
+    private final Icon icon = AppIcons.TOOL_WINDOW_OBSERVABILITY;
+    private final Icon iconWithGreenDot = ExecutionUtil.getLiveIndicator(icon);
+    private final String localHostname;
     private final Project project;
-    private JBCefBrowser jbCefBrowser;
+    private final AnalyticsService analyticsService;
 
     //the recent activity code is not managed in one place that is accessible from the plugin code
     // like a project service, so currently need an init task.
     // it is used to send live data in case the live view button was clicked before the recent activity window was initialized.
-    private MyInitTask initTask = null;
+    private MyInitTask initTask;
+    private Timer myLiveDataTimer;
+    private Timer activityFetchingTimer;
+    private JBCefBrowser jbCefBrowser;
+    private RecentActivityResult latestActivityResult;
+    private boolean webAppInitialized;
 
-    private Timer myLiveDataTimer = null;
+    public static RecentActivityService getInstance(Project project) {
+        return project.getService(RecentActivityService.class);
+    }
 
     public RecentActivityService(Project project) {
         this.project = project;
+        //initialize AnalyticsService early so the UI already can detect the connection status when created
+        this.analyticsService = project.getService(AnalyticsService.class);
+        this.localHostname = CommonUtils.getLocalHostname();
+        this.latestActivityResult = new RecentActivityResult(null, new ArrayList<>());
     }
 
+    public void startFetchingActivities() {
+        if (activityFetchingTimer != null)
+            return;
 
-    public static RecentActivityService getInstance(Project project){
-        return project.getService(RecentActivityService.class);
+        var activityFetchingTask = new TimerTask() {
+            @Override
+            public void run() {
+                fetchRecentActivities();
+            }
+        };
+        activityFetchingTimer = new Timer();
+        activityFetchingTimer.scheduleAtFixedRate(activityFetchingTask, 0, FETCHING_LOOP_INTERVAL);
+    }
+
+    public Content createTabContent() {
+        if (!JBCefApp.isSupported()) {
+            // Fallback to an alternative browser-less solution
+            return null;
+        }
+
+        var contentFactory = ContentFactory.getInstance();
+        var editorService = project.getService(EditorService.class);
+        var customViewerWindow = new CustomViewerWindow(project, RESOURCE_FOLDER_NAME, new HashMap<>() {{
+            put(RECENT_EXPIRATION_LIMIT_VARIABLE, RECENT_EXPIRATION_LIMIT_MILLIS);
+        }});
+        jbCefBrowser = customViewerWindow.getWebView();
```

Comment:
CustomViewerWindow is a completely redundant class that has no justification for living. it is just a one to one copy from some old scala example in the internet.
and worst, CustomViewerWindow hold a reference to the jcefbrowser in the member webView, it will probably be released because its a local variable here but the creation of the jcefbrowser is a few lines of code that should be moved to this class,RecentActivityService

Reply:
and more on that, CustomSchemeHandlerFactory , CustomResourceHandlerUtil, CustomResourceHandler,
are also a one to one copy from the same scala example, they are too complex and hard to understand.
see for example JaegerUISchemeHandlerFactory and JaegerUIResourceHandler that do the same job but are much simpler.
IMO all these copied classes should be refactrored and we can create a common resource loader for all our jcef browsers. but that can be done in another task, maybe you can open an issue about that.

Reply:
Removed the usage of CustomViewerWindow
I'll open a task about  CustomSchemeHandlerFactory , CustomResourceHandlerUtil, CustomResourceHandler